File: src/main/java/uk/edu/populationfitness/models/genes/sphere/SchumerSteiglitzGenes.java
package uk.edu.populationfitness.models.genes.sphere;

import uk.edu.populationfitness.models.Config;
import uk.edu.populationfitness.models.fastmaths.ExpensiveCalculatedValues;
import uk.edu.populationfitness.models.fastmaths.FastMaths;
import uk.edu.populationfitness.models.fastmaths.ValueCalculator;
import uk.edu.populationfitness.models.genes.bitset.NormalizingBitSetGenes;

public class SchumerSteiglitzGenes extends NormalizingBitSetGenes {
    private static class NormalizationRatioCalculator implements ValueCalculator<Double> {
        @Override
        public Double calculateValue(long n) {
            return 100000000.0 * n;
        }
    }

    private static final ExpensiveCalculatedValues<Double> NormalizationRatios = new ExpensiveCalculatedValues(new SchumerSteiglitzGenes.NormalizationRatioCalculator());

    public SchumerSteiglitzGenes(Config config) {
        super(config, 100.0);
    }

    @Override
    protected double calculateNormalizationRatio(int n) {
        return NormalizationRatios.findOrCalculate(n);
    }

    @Override
    protected double calculateFitnessFromGenes(double[] unknowns) {
        /*
          f left (x right ) = sum from {i=1} to {n} {{x} rsub {i} rsup {4}}
         */

        double fitness = 0.0;

        for (double x : unknowns) {
            fitness += FastMaths.pow(x, 4);
        }

        return fitness;
    }
}

File: DotNet/PopulationFitness/PopulationFitness/Models/Genes/Sphere/SchumerSteiglitzGenes.cs
using PopulationFitness.Models.FastMaths;
using PopulationFitness.Models.Genes.BitSet;

namespace PopulationFitness.Models.Genes.Sphere
{
    public class SchumerSteiglitzGenes : NormalizingBitSetGenes
    {
        private class NormalizationRatioCalculator : IValueCalculator<double>
        {

            public double CalculateValue(long n)
            {
                return 100000000.0 * n;
            }
        }

        private static readonly ExpensiveCalculatedValues<double> NormalizationRatios = new ExpensiveCalculatedValues<double>(new SchumerSteiglitzGenes.NormalizationRatioCalculator());

        public SchumerSteiglitzGenes(Config config) : base(config, 100.0)
        { }

        protected override double CalculateNormalizationRatio(int n)
        {
            return NormalizationRatios.FindOrCalculate(n);
        }

        protected override double CalculateFitnessFromIntegers(long[] integer_values)
        {
            /*
              f left (x right ) = sum from {i=1} to {n} {{x} rsub {i} rsup {4}}
             */

            double fitness = 0.0;

            foreach (long integer_value in integer_values)
            {
                double x = Interpolate(integer_value);

                fitness += FastMaths.FastMaths.Pow(x, 4);
            }

            return fitness;
        }
    }
}
